overlord/snapstate/snapmgr.go

```diff
@@ -682,9 +684,21 @@ func (m *SnapManager) StartUp() error {
 		return fmt.Errorf("failed to generate cookies: %q", err)
 	}
 
+	m.changeCallbackID = m.state.AddChangeStatusChangedHandler(processInhibitedAutoRefresh)
```

Comment:
this merits a comment about what is being done here? also can you explain why you chose to set up things this way and have explicit emittting in a few places?

Reply:
The only other entrypoint I can think of that can guarantee that all snaps were checked is the `check-rerefresh` handler, it would be something like this:
```diff
diff --git a/overlord/snapstate/handlers.go b/overlord/snapstate/handlers.go
index c49079088b..5e391f43ad 100644
--- a/overlord/snapstate/handlers.go
+++ b/overlord/snapstate/handlers.go
@@ -4356,6 +4356,11 @@ func (m *SnapManager) doCheckReRefresh(t *state.Task, tomb *tomb.Tomb) error {
                return &state.Retry{After: reRefreshRetryTimeout, Reason: "pending refreshes"}
        }
 
+       chg := t.Change()
+       if chg.Kind() == "auto-refresh" {
+               defer maybeAddRefreshInhibitNotice(st)
+       }
+
        snaps, failed, err := refreshedSnaps(t)
        if err != nil {
                return err
@@ -4366,8 +4371,6 @@ func (m *SnapManager) doCheckReRefresh(t *state.Task, tomb *tomb.Tomb) error {
                }
        }
 
-       chg := t.Change()
-
        // if any snap failed to refresh, reconsider validation set tracking
        if failed {
                tasksets, err := maybeRestoreValidationSetsAndRevertSnaps(st, snaps, chg.ID())
```

but we would still need the pre-dl check in `launchAutoRefresh` for the case that all snaps are inhibited early in the soft check.

The main requirement for the emitting is being able to observe the overall auto-refresh change status for all snaps. The change status change handler seemed like a natural place for this.

I don't have strong preferences here, would love to know if you have other entrypoints in mind?

Reply:
Added a comment explaining what was being registered and why, thanks!